### PR TITLE
Test gen support string additional options 2

### PR DESCRIPTION
**Pull request by allredj**
This replaces #695

This adds the option --string-max-length to the string solver, this constrains all the strings to be smaller than the maximal length. This shouldn't affect much the performance of the decision procedure but may be necessary for obtaining concrete traces (when the --trace option is activated).

I also added the options --string-non-empty to constrain the nondet strings to have length at least 1 (to avoid obtaining counter-example that concatenates the empty string two times, which is not very interesting), and --string-printable to constrain strings to contain only characters in the range between (space, U+0020) and ~ (tilde, U+007E). making the example nicer than a sequence of null characters.

This also fixes a couple of bugs in the solver.
This corresponds to PR #652 on master

### CODE REVIEW

File: src/cbmc/cbmc_parse_options.h [hunk 1]
```diff
@@ -37,7 +37,7 @@ class optionst;
   "(no-sat-preprocessor)" \
   "(no-pretty-names)(beautify)" \
   "(dimacs)(refine)(max-node-refinement):(refine-arrays)(refine-arithmetic)"\
-  "(refine-strings)" \
+  "(refine-strings)(string-non-empty)(string-printable)(string-max-length):" \
```
Comment (peterschrammel) [inline src/cbmc/cbmc_parse_options.h]: put options on separate lines to make merging easier

File: src/solvers/refinement/string_constraint_generator.h [hunk 2]
```diff
@@ -26,9 +26,17 @@ class string_constraint_generatort
   // to the axiom list.
 
   string_constraint_generatort():
-    mode(ID_unknown)
+    mode(ID_unknown),
+    max_string_length(-1),
```
Comment (peterschrammel) [inline src/solvers/refinement/string_constraint_generator.h]: Is this convention documented?

File: src/solvers/refinement/string_constraint_generator.h [hunk 3]
```diff
@@ -26,9 +26,17 @@ class string_constraint_generatort
...
+    force_printable_characters(false)
   { }
 
+  // Constraints on the maximal length of strings
+  int max_string_length;
```
Comment (peterschrammel) [inline src/solvers/refinement/string_constraint_generator.h]: Hm... here we have an int now?

**Comment by allredj** [comment]
Ready for review in: https://github.com/diffblue/cbmc/pull/699

File: src/solvers/refinement/string_refinement.cpp [hunk 4]
```diff
@@ -283,52 +315,73 @@ bool string_refinementt::add_axioms_for_string_assigns(const exprt &lhs,
...
     {
-      string_exprt str=to_string_expr(it.second);
-      exprt length=current_model[str.length()];
-      exprt content=str.content();
-      replace_expr(symbol_resolve, content);
-      found_length[content]=length;
-      mp_integer found_length;
-      if(!to_integer(length, found_length))
+      assert(found_length.is_long());
+      if(found_length < 0)
```
Comment (peterschrammel) [inline src/solvers/refinement/string_refinement.cpp]: no spaces

File: src/solvers/refinement/string_refinement.h [hunk 5]
```diff
@@ -38,8 +33,13 @@ class string_refinementt: public bv_refinementt
   // Should we use counter examples at each iteration?
   bool use_counter_example;
 
+  // Should we concretize strings when the solver finished
   bool do_concretizing;
 
+  void set_max_string_length(unsigned int i);
```
Comment (peterschrammel) [inline src/solvers/refinement/string_refinement.h]: size_t?

**Comment by allredj** [comment]
Requested fixes have been applied and squashed.

**Pull request merged**